bridge-token/src/lib.rs

```diff
@@ -67,14 +71,18 @@ impl BridgeToken {
         reference: Option<String>,
         reference_hash: Option<Base64VecU8>,
         decimals: Option<u8>,
+        icon: Option<String>,
     ) {
         // Only owner can change the metadata
-        assert_self();
+        assert!(self.controller_or_self());
+
         name.map(|name| self.name = name);
         symbol.map(|symbol| self.symbol = symbol);
         reference.map(|reference| self.reference = reference);
         reference_hash.map(|reference_hash| self.reference_hash = reference_hash);
         decimals.map(|decimals| self.decimals = decimals);
+        #[cfg(feature = "migrate_icon")]
+        icon.map(|icon| self.icon = Some(icon));
     }
```

Comment:
if we accept a function argument without the feature enabled, then we need to apply it without the feature as well. Or explicitly specify that we ignore that and don't allow making such a call with the provided icon. Otherwise, it's misleading

Reply:
I tried doing this:

```rust
pub fn set_metadata(
        &mut self,
        name: Option<String>,
        symbol: Option<String>,
        reference: Option<String>,
        reference_hash: Option<Base64VecU8>,
        decimals: Option<u8>,
        #[cfg(feature = "migrate_icon")] icon: Option<String>,
    )
```

But it doesn't work

```
error[E0061]: this function takes 5 arguments but 6 arguments were supplied
  --> src/lib.rs:67:12
   |
67 |       pub fn set_metadata(
   |       -      ^^^^^^^^^^^^ expected 5 arguments
   |  _____|
   | |
68 | |         &mut self,
69 | |         name: Option<String>,
   | |         ----
70 | |         symbol: Option<String>,
   | |         ------
71 | |         reference: Option<String>,
   | |         ---------
72 | |         reference_hash: Option<Base64VecU8>,
   | |         --------------
73 | |         decimals: Option<u8>,
   | |         --------
74 | |         #[cfg(feature = "migrate_icon")] icon: Option<String>,
   | |                                          ---- supplied 6 arguments
75 | |     ) {
   | |_____- defined here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0061`.
error: could not compile `bridge-token`.

To learn more, run the command again with --verbose.
make: *** [res/bridge_token_no_icon.wasm] Error 101
```

The other option I see is having two different functions, but didn't liked that approach. Are you suggesting to have the two functions, one for "migrate_icon" feature, and one without it? I think it is ok to have the function receiving the icon and not using it. The main reason for me to add this feature flag is for testing purposes.

Reply:
Yes, I think two different functions could work but I don't like the approach, seems for me like a redundant overhead. Let's just add at least a log that if the icon is provided and `migrate_icon` feature is not enabled, let's print a log that the icon is being ignored.

Reply:
added log message.